src/lib/backoff.ts

```diff
@@ -0,0 +1,101 @@
+import { strict as assert } from 'assert';
+import { promisify } from 'util';
+
+import { StatusError } from './errors';
+
+// TODO use standard lib async setTimout (requires node 16)
+const sleep = promisify(setTimeout);
+
+export type OnRetryOpts = {
+	retryCount: number;
+	delay: number;
+	error: any;
+};
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (info: OnRetryOpts) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: Infinity,
+};
+
+/**
+ * Retries a function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options as a way for callers to react to retries
+ */
+export function withBackoff<T extends (...args: any[]) => any>(
+	func: T,
+	options: Partial<Options> = {},
+) {
+	assert(typeof func === 'function', 'expected a function as parameter');
+	// If the function returns a promise, unwrap the promise return type
+	// otherwise use the actual return
+	type TReturn = ReturnType<T> extends Promise<infer R> ? R : ReturnType<T>;
+
+	const normalizedOptions: Options = {
+		...DEFAULT_OPTIONS,
+		...options,
+	} as Options;
+
+	return async function wrapped(...args: Parameters<T>): Promise<TReturn> {
+		try {
+			return await func(...args);
+		} catch (error) {
+			if (isRetryError(error)) {
+				normalizedOptions.attempt = 0;
```

Comment:
Why reset the number of attempts on this case? Even though you are using a different backoff time, having the right number of attempts will still be useful.

Reply:
I see you answered this above. Even though I see your point, I’m still not sure about the reset, since that might result in a weird interface when using the `retryCallback`, since the value passed for the attempts opt is non monotonic.

Reply:
I removed the attempt = 0 since the maxDelay will prevent the big attempt number from causing the device to wait a long time. Additionally, as we spoke in our check-in, it's weird if an observer of the onRetry callback logs attempt = 0 over and over again even though it is incrementing...